zebra-consensus/src/transaction/tests.rs

```diff
@@ -62,19 +37,82 @@ fn v5_fake_transactions_for_network(network: Network) -> Result<(), Report> {
                     ..
                 } => {
                     if let Some(s) = sapling_shielded_data {
-                        super::check::sapling_balances_match(&s)?;
+                        check::sapling_balances_match(&s)?;
 
                         for spend in s.spends_per_anchor() {
-                            super::check::spend_cv_rk_not_small_order(&spend)?
+                            check::spend_cv_rk_not_small_order(&spend)?
                         }
                         for output in s.outputs() {
-                            super::check::output_cv_epk_not_small_order(&output)?;
+                            check::output_cv_epk_not_small_order(&output)?;
                         }
                     }
                 }
                 _ => panic!("we should have no tx other than 5"),
             }
         }
     }
+
     Ok(())
 }
+
+#[test]
+fn v5_transaction_with_no_inputs_fails_validation() {
+    let transaction = v5_fake_transactions_for_network(
+        Network::Mainnet,
+        zebra_test::vectors::MAINNET_BLOCKS.iter(),
+    )
+    .rev()
+    .find(|transaction| {
+        transaction.inputs().is_empty()
+            && transaction.sapling_spends_per_anchor().next().is_none()
+            && transaction.orchard_actions().next().is_none()
+            && transaction.joinsplit_count() == 0
+            && (!transaction.outputs().is_empty() || transaction.sapling_outputs().next().is_some())
+    })
+    .expect("At least one transaction with no inputs in the test vectors");
+
+    assert_eq!(
+        check::has_inputs_and_outputs(&transaction),
+        Err(TransactionError::NoInputs)
+    );
+}
+
+#[test]
+fn v5_transaction_with_no_outputs_fails_validation() {
+    let transaction = v5_fake_transactions_for_network(
+        Network::Mainnet,
+        zebra_test::vectors::MAINNET_BLOCKS.iter(),
+    )
+    .rev()
+    .find(|transaction| {
+        transaction.outputs().is_empty()
+            && transaction.sapling_outputs().next().is_none()
+            && transaction.orchard_actions().next().is_none()
+            && transaction.joinsplit_count() == 0
+            && (!transaction.inputs().is_empty()
+                || transaction.sapling_spends_per_anchor().next().is_some())
+    })
+    .expect("At least one transaction with no outputs in the test vectors");
+
+    assert_eq!(
+        check::has_inputs_and_outputs(&transaction),
+        Err(TransactionError::NoOutputs)
+    );
+}
+
+fn v5_fake_transactions_for_network<'b>(
```

Comment:
Can you move this function to the `arbitrary` module, so it can be used by other Zebra crates?

(Test modules don't get exposed to other crates' tests. We use a `proptest-impl` feature to work around that.)